arelle/plugin/validate/DBA/rules/fr.py

```diff
@@ -797,6 +797,55 @@ def rule_fr81(
         )
 
 
+@validation(
+    hook=ValidationHook.XBRL_FINALLY,
+)
+def rule_fr89(
+        pluginData: PluginValidationDataExtension,
+        val: ValidateXbrl,
+        *args: Any,
+        **kwargs: Any,
+) -> Iterable[Validation]:
+    """
+    DBA.FR89: If ClassOfReportingEntity is one of the following values:
+
+    Regnskabsklasse C, mellemstor virksomhed // Reporting class C, medium-size enterprise
+    Regnskabsklasse C, stor virksomhed // Reporting class C, large enterprise
+    Regnskabsklasse D // Reporting class D
+    Then TypeOfAuditorAssistance should be: Revisionspåtegning // Auditor's report on audited financial statements
+    """
+    modelXbrl = val.modelXbrl
+    facts = modelXbrl.factsByQname.get(pluginData.classOfReportingEntityQn)
+    if facts is not None:
+        for fact in facts:
+            if fact.xValid >= VALID:
+                if fact.xValue in [
+                    pluginData.reportingClassCLargeDanish,
+                    pluginData.reportingClassCLargeEnglish,
+                    pluginData.reportingClassCMediumDanish,
+                    pluginData.reportingClassCMediumEnglish,
+                    pluginData.reportingClassDDanish,
+                    pluginData.reportingClassDEnglish,
+                ]:
+                    auditorFacts = modelXbrl.factsByQname.get(pluginData.typeOfAuditorAssistanceQn)
+                    if auditorFacts is not None:
+                        for auditorFact in auditorFacts:
+                            if auditorFact.xValid >= VALID and auditorFact.xValue not in [
+                                pluginData.auditedFinancialStatementsDanish,
+                                pluginData.auditedFinancialStatementsEnglish
+                            ]:
+                                yield Validation.error(
```

Comment:
This error could potentially be raised multiple times if there are multiple classOfReportingEntity or typeOfAuditorAssistance facts. Is that the intended behavior?

Additionally, no error is raised when there are matching classOfReportingEntity values but no valid auditor facts. That seems like an oversight.

Note that some work within each iteration of the loop, such as building a list of reporting class values and performing auditor fact lookups, is not dependent on the iteration and could be done once outside the loop for efficiency.